external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenSparkConf.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.kafka010
+
+import org.apache.kafka.common.security.auth.SecurityProtocol.SASL_SSL
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.Kafka
+
+private[spark] case class KafkaTokenClusterConf(
+  identifier: String,
+  bootstrapServers: String,
+  securityProtocol: String,
+  kerberosServiceName: String,
+  trustStoreLocation: Option[String],
+  trustStorePassword: Option[String],
+  keyStoreLocation: Option[String],
+  keyStorePassword: Option[String],
+  keyPassword: Option[String],
+  tokenMechanism: String) {
+  override def toString: String = s"KafkaTokenClusterConf{"
+    s"identifier=$identifier, " +
+    s"bootstrapServers=$bootstrapServers, "
+    s"securityProtocol=$securityProtocol, " +
+    s"kerberosServiceName=$kerberosServiceName, " +
+    s"trustStoreLocation=$trustStoreLocation, " +
+    s"trustStorePassword=${trustStorePassword.map(_ => "xxx")}, " +
+    s"keyStoreLocation=$keyStoreLocation, "
+    s"keyStorePassword=${keyStorePassword.map(_ => "xxx")}, " +
+    s"keyPassword=${keyPassword.map(_ => "xxx")}, " +
+    s"tokenMechanism=$tokenMechanism}"
+}
+
+private[spark] class KafkaTokenSparkConf(sparkConf: SparkConf) extends Logging {
+  def getClusterIdentifiers(): Array[String] = {
+    val result = sparkConf.get(Kafka.CLUSTERS) match {
```

Comment:
You could use e.g. `sparkConf.getAllWithPrefix("spark.kafka.clusters.")` and avoid having to list cluster names in this config. (You could parse the cluster names from the keys in the returned list.)

Or if you want to keep this, it should be a sequence config (`.toSequence`).

But given the whole change seems to be about exposing the cluster name in a bunch of places, I'm wondering if there isn't a cleaner way to do that?

e.g. don't you have the server name you're connecting to, and then you could check it against the "bootstrapServers" of each cluster-specific config to find the one you're looking for?

Reply:
`getAllWithPrefix` is definitely a good idea and will remove one ugly indirection. Adapting it...

Related the cluster name exposure I've considered the mentioned solution but didn't find it stable enough because of the following. `bootstrapServers` will work only if the same servers configured for delegation token and data processing. Here is an example:

Servers: `host1, host2`

Working scenario:
* Token `bootstrapServers`: `host1:9093, host2:9093`
* Data processing `bootstrapServers`: `host1:9093, host2:9093`

Valid but not working scenario:
* Token `bootstrapServers`: `host1:9093`
* Data processing `bootstrapServers`: `host2:9093`

The second case is valid because as the name tells the server list is only a bootstrap and when the first server in the list is reachable the Kafka client discovers the remaining hosts internally.

To resolve this I've tried the following:
* Was thinking about to check whether at least 1 host matches but I think it's brittle
* Checked the Kafka API but until now haven't found any `resolveBootstrapServers` or something similar.
* Thinking about to put some constraint on the `bootstrapServers` to be the same but without resolution don't see it's feasible

Thinking about it further...

Reply:
Since only the `getAllWithPrefix` solved leave it open.

Reply:
Can the Kafka libraries figure out the right delegation token when talking to a remote server?

e.g., in HDFS, if your app talks to multiple NNs, you'll get tokens for all of them, and the HDFS client library knows to pick the right token when trying to connect to a specific NN.

Can the Kafka libraries do that?

Otherwise this feels incredibly brittle. You're basically forcing the user to write the correct configuration and bake the right cluster ID into their app code. That feels to me like it would lead to very confused users when things don't work.

Reply:
Also, if Kafka libraries don't handle that by themselves, I think a better approach would be to have the configs you're adding be sort of a "Kafka topology map". e.g. they also contain which kafka brokers belong to which "cluster".

That way, at least, an admin can create the default Spark config and users don't need to touch their application code.

Reply:
> Can the Kafka libraries figure out the right delegation token when talking to a remote server?

That's a really good idea and was excited until I've figured out that it's not implemented in Kafka :/
Login modules like `Krb5LoginModule` and `ScramLoginModule` support different authentication control flags like `required`, etc... When `sufficient` provided then the  LoginModule is not required to succeed and if it fails, authentication continues down the LoginModule list. My plan was to add `ScramLoginModule` as many times as the number of available delegation tokens with `sufficient` flag.

I've tested it but its not implemented in Kafka ([this](https://github.com/apache/kafka/blob/02221bd907a23041c95ce6446986bff631652b3a/clients/src/main/java/org/apache/kafka/common/security/JaasContext.java#L95) code blows up):
```
Exception in thread "main" org.apache.kafka.common.KafkaException: Failed create new KafkaAdminClient
	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:370)
	at org.apache.kafka.clients.admin.AdminClient.create(AdminClient.java:52)
	at com.kafka.delegationtoken.consumer.SecureKafkaConsumer$.main(SecureKafkaConsumer.scala:96)
	at com.kafka.delegationtoken.consumer.SecureKafkaConsumer.main(SecureKafkaConsumer.scala)
Caused by: java.lang.IllegalArgumentException: JAAS config property contains 2 login modules, should be 1 module
	at org.apache.kafka.common.security.JaasContext.load(JaasContext.java:95)
	at org.apache.kafka.common.security.JaasContext.loadClientContext(JaasContext.java:84)
	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:119)
	at org.apache.kafka.common.network.ChannelBuilders.clientChannelBuilder(ChannelBuilders.java:65)
	at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:88)
	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:346)
	... 3 more
```
I've filed [KAFKA-8234](https://issues.apache.org/jira/browse/KAFKA-8234) but I think its not an option for this case (even if its implemented not all Kafka brokers will have it).

What I'm thinking of is to take over the `bootstrapServers` parameter from token side if cluster identifier given. The logic would be something like:
* Use `clusterId` parameter name instead of `tokenClusterId`
* If `bootstrapServers` provided only on token side and `clusterId` defined on processing side then take the servers over (only one place of configuration).
* If `bootstrapServers` provided both on token + data processing side then give a warning (maybe user don't know it can be taken over or just accidentally misconfigured). This can be even an exception which doesn't let the query start. I'm always afraid of introducing something like this but personally think the second is the right way.
* If `bootstrapServers` provided only on data processing side then work just like before

Reply:
It seems you're still proposing a change that requires a new user-level API for this that ties a broker to a "cluster ID". (Either an explicit parameter or a new config, but either way, it requires changes to the application code.)

I don't think that's a very good idea. It requires changes in user's applications for it to work, which makes it hard to run the same application against different clusters, for example. I think if you could do the mapping entirely in the config, and let the application just define the broker it wants to use, that's much much better.

Reply:
Since I ran out of ideas I thought there is a tradeoff between having redundant `bootstrapServers` configuration vs introducing new API for a new feature I considered the second is a better option.

I see your point about the one-time significant cost to adapt all the applications which can make upgrade an expensive operation for heavy users with lot of apps. It makes sense.

Without API change I see only this option left:
* `bootstrapServers` configured on token side (admin task)
* Spark obtains and stores  tokens together with the configured `bootstrapServers`
* `bootstrapServers` configured on app side (app dev task and this case existing apps don't have to be touched)
* When token needed Spark checks every stored tokens `bootstrapServers` and if there is an intersection with the app side configuration then it will use it

This approach makes an assumption that there is an intersection between the 2 `bootstrapServers` configs as we've discussed previously.

Please elaborate if you thought something better when you've written this
> I think if you could do the mapping entirely in the config, and let the application just define the broker it wants to use, that's much much better.

Reply:
Had a deeper look at [KAFKA-8234](https://issues.apache.org/jira/browse/KAFKA-8234) with the Kafka guys and the good part its most probably client code modification which makes this a possible solution for long term (if it's implemented). The bad part it would be a huge change with several questions which makes it not an option for now.

Independently from the approach chosen here I've put this on my backlog.

Reply:
I explained my suggestion previously. You need a config to list the brokers associated with a bootstrap server in your "cluster" definition. Then you can match the tokens you create to the broken you're trying to contact based on that mapping.

Reply:
I think I see your point.

Each and every node can be contacted as bootstrap server in a cluster and it will find all the remaining brokers. So if there are 100 brokers list all the nodes in a config with the remaining 99 brokers?
```
spark.kafka.clusterdefinition.host1:port1.brokers ”host2:port1,host3:port1...”
spark.kafka.clusterdefinition.host2:port1.brokers ”host1:port1,host3:port1...”
...
```

The other way would be this (instead of `bootstrap.servers` `allBrokers` used):
```
spark.kafka.clusters.cluster1.allBrokers ”host1:port1,host2:port1”
spark.kafka.clusters.cluster1…
```

If we choose this way I think what we have to achieve is to have intersection between `bootstrapServers` on data processing side and `bootstrapServers` on the token side. Since this way its configured 2 places hard to enforce (though the probability is really low to misconfigure it).

Reply:
The only way I can see to avoid the user providing this information in their application is to have it in the configuration. Or to have Kafka figure it out by itself.

I prefer not to have users change their application, because that *will* get messy. Changing a configuration file is much easier (and can be automated).

Reply:
I think such configuration can be added:
```
spark.kafka.clusters.cluster1.targetBrokers ”*.company.com:port1”
```
This configuration would set the target where the obtained delegation token should be used (namely all sources/sinks will use it where `bootstrapServers` has match). Adapting the PR...

Reply:
Solving this since new approach added.